### PR TITLE
Pmrozik/material information

### PR DESCRIPTION
**Pull request by PiotrMrozik**

### CODE REVIEW

File: test/src/textureTest.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+TEST_F(TextureTest, rgl_texture_invalid_argument)
+{
+	rgl_texture_t texture= nullptr;
+	float* textureRawData = generateStaticColorTexture(100, 100, 100.0f);
+
+	EXPECT_RGL_INVALID_ARGUMENT(rgl_texture_create(nullptr, textureRawData, 100, 100), "texture != nullptr");
+	EXPECT_RGL_INVALID_ARGUMENT(rgl_texture_create(&texture, nullptr, 100, 100), "texels != nullptr");
+	EXPECT_RGL_INVALID_ARGUMENT(rgl_texture_create(&texture, textureRawData, -1, 100), "width > 0");
+	EXPECT_RGL_INVALID_ARGUMENT(rgl_texture_create(&texture, textureRawData, 100, 0), "height > 0");
```
Comment (msz-rai) [inline test/src/textureTest.cpp:29]: Arguments between calls are not re-initialized.

File: test/src/textureTest.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+{
+	auto [width, height, value] = GetParam();
+
+	rgl_texture_t texture= nullptr;
+	rgl_entity_t entity = nullptr;
+	rgl_mesh_t mesh = makeCubeMesh();
+	auto* textureRawData = generateStaticColorTexture(width, height, value);
+
+	EXPECT_RGL_SUCCESS(rgl_texture_create(&texture, textureRawData, width, height));
+	EXPECT_RGL_SUCCESS(rgl_mesh_set_texture_coords(mesh, cubeUVs, 8));
```
Comment (msz-rai) [inline test/src/textureTest.cpp:44]:
```suggestion
	EXPECT_RGL_SUCCESS(rgl_mesh_set_texture_coords(mesh, cubeUVs, ARRAY_SIZE(cubeUVs)));
```

File: test/src/textureTest.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	for (int i = 0; i < outCount; ++i)
+	{
+		EXPECT_NEAR(value, outIntensity.at(i), EPSILON_F);
+	}
+
+	delete[] textureRawData;
+}
+
+
+TEST_P(TextureTest, rgl_texture_use_case)
```
Comment (msz-rai) [inline test/src/textureTest.cpp:88]: Please add a comment before the test to explain what is going on here, what to expect, and how to check if the test passes.

File: test/src/textureTest.cpp
Comment (msz-rai) [inline test/src/textureTest.cpp:1]: I'd move this test to src/scene/TextureTest.cpp

**Review by msz-rai — CHANGES_REQUESTED** [review]
Some changes from previous review have been applied, but not all of them. Please take a look at the new comments also.

File: test/include/utils.hpp [hunk 4]
```diff
@@ -173,3 +174,46 @@ static rgl_mesh_t loadMesh(std::filesystem::path path)
 	EXPECT_RGL_SUCCESS(rgl_mesh_create(&mesh, vs.data(), vs.size(), is.data(), is.size()));
 	return mesh;
 }
+
+template<typename T>
+static T* generateStaticColorTexture(int width, int height, T value)
```
[the remark this answers is not in the document]
Comment (PiotrMrozik) [reply test/include/utils.hpp:179]: I will leave the template. For now it changes nothing, but other branch will present this feature alongside with multi-channel, multi-type texture. This is future-proof.

File: test/src/tapeSurfaceTest.cpp [hunk 5]
```diff
@@ -38,6 +38,15 @@ TEST_F(TapeCase, RecordPlayAllCalls)
 	EXPECT_RGL_SUCCESS(rgl_entity_create(&entity, nullptr, mesh));
 	EXPECT_RGL_SUCCESS(rgl_entity_set_pose(entity, &identityTf));
 
+	rgl_texture_t texture = nullptr;
+	int width = 1024;
+	int height = 2048;
+	auto textureRawData = generateCheckerboardTexture<char>(width, height);
```
Comment (PiotrMrozik) [inline test/src/tapeSurfaceTest.cpp:44]:
```suggestion
	auto textureRawData = generateCheckerboardTexture<unsigned char>(width, height);
```

File: test/include/utils.hpp [hunk 6]
```diff
@@ -173,3 +174,46 @@ static rgl_mesh_t loadMesh(std::filesystem::path path)
...
+		}
+		yStep = 0;
+		if (i % xGridSize == 0)
+		{
+			xStep += xGridSize;
+		}
+	}
+
+	return texels;
+}
```
[the remark this answers is not in the document]
Comment (PiotrMrozik) [reply test/include/utils.hpp]: When we will make bigger cleanup for test structures, full unite coverage, and other namespaces. Not now. Leave it for better times.

File: src/scene/Texture.cpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+void Texture::createTextureObject(const void* texels, int width, int height)
+{
+	cudaResourceDesc res_desc = {};
+
+	int32_t numComponents = 1;
+
+	cudaChannelFormatDesc channel_desc = cudaCreateChannelDesc<uchar1>();
+
+	int32_t pitch = width * numComponents * sizeof(unsigned char);
```
Comment (msz-rai) [inline src/scene/Texture.cpp:38]: I don't like that the type of texels is repeated in multiple places (here, `optixPrograms.cu`, `apiCore.cpp`, in tests). I'd define a separate type for it. It has already caused a bug pointed out in this comment.

File: src/scene/Texture.cpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+Texture::~Texture()
+{
+	cleanup();
+}
+
+void Texture::cleanup()
+{
+	cudaDestroyTextureObject(dTextureObject);
+	cudaFreeArray(dPixelArray);
```
Comment (msz-rai) [inline src/scene/Texture.cpp:79]: What if we pass `nullptr` here? I'd free only when `dPixelArray` is not `nullptr`.

File: test/src/scene/textureTest.cpp [hunk 9]
```diff
@@ -0,0 +1,152 @@
+#include <gtest/gtest.h>
+#include "utils.hpp"
+#include "scenes.hpp"
+#include "lidars.hpp"
+
+#ifdef RGL_BUILD_ROS2_EXTENSION
+#include <rgl/api/extensions/ros2.h>
+#endif
+
+constexpr unsigned short UsMaxValue = 255;
```
Comment (msz-rai) [inline test/src/scene/textureTest.cpp:10]: `std::numeric_limits<unsigned short>::max()` should work (with `#include <limits>`)

**Review by msz-rai — COMMENTED** [review]
:+1:
Please address the comments.

File: test/src/scene/textureTest.cpp (hunk 9 above)
Comment (PiotrMrozik) [reply test/src/scene/textureTest.cpp:10]: Discussed offline

File: include/rgl/api/core.h [hunk 10]
```diff
@@ -339,6 +364,35 @@ rgl_entity_destroy(rgl_entity_t entity);
...
+/**
+ * Creates a Texture.
+ * Texture is a container object which holds device pointer to texture resource.
+ * @param out_texture Handle to the created Texture.
+ * @param texels Pointer to the texture data.
+ * @param width Width of the texture. Has to be positive.
+ * @param height Height of the texture. It is not demanded that width == height. Has to be positive.
+ */
+RGL_API rgl_status_t
+rgl_texture_create(rgl_texture_t* out_texture, const void* texels, int32_t width, int32_t height);
```
[the remark this answers is not in the document]
Comment (PiotrMrozik) [reply include/rgl/api/core.h:386]: Added documentation.
I would rather leave it as void* for easier further development of multi-channel, multi-type texture passing.

**Pull request merged**